### PR TITLE
Adding Gen pipeline options

### PR DESCRIPTION
**Pull request by adstraw**
Usage e.g.  PLAIDML_TARGET="intel_gen_ocl_spirv{use-block-ops=false spirv-version=120}"

### CODE REVIEW

File: pmlc/target/intel_gen_ocl_spirv/spirv_target.cc [hunk 1]
```diff
@@ -15,12 +15,20 @@ namespace {
...
+  IntelGenOclAddSpirvTarget() = default;
+  explicit IntelGenOclAddSpirvTarget(unsigned spirvVersion) {
+    this->spirvVersion = spirvVersion;
+  }
   void runOnOperation() {
     auto target_env = getOperation().getAttrOfType<spirv::TargetEnvAttr>(
         spirv::getTargetEnvAttrName());
     if (!target_env) {
+      auto version = spirv::Version::V_1_5;
+      if (spirvVersion == 120) {
```
Comment (abialas1) [inline pmlc/target/intel_gen_ocl_spirv/spirv_target.cc:27]: Do we want to fix on 1.5 and 1.2 only or make it more generic?

File: pmlc/target/x86/register.cc [hunk 2]
```diff
@@ -20,7 +20,10 @@ static PassPipelineRegistration<> passPipelineReg(kPassPipelineTargetName,
 
 class Target : public compiler::Target {
 public:
-  void buildPipeline(mlir::OpPassManager &pm) { pipelineBuilder(pm); }
+  void buildPipeline(mlir::OpPassManager &pm, llvm::StringRef targetOptions) {
+    assert(targetOptions.empty());
```
Comment (abialas1) [inline pmlc/target/x86/register.cc:24]: Is this assert needed?

**Comment by abialas1** [comment]
Looks good to me, added some minor comments

File: pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc [hunk 3]
```diff
@@ -31,6 +31,22 @@ struct StdxSubgroupBroadcastOpConversion final
     : public SPIRVOpLowering<stdx::SubgroupBroadcastOp> {
   using SPIRVOpLowering<stdx::SubgroupBroadcastOp>::SPIRVOpLowering;
 
+  LogicalResult
```
Comment (jbruestle) [inline pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc:34]: Maybe template on the op to generate to reduce code duplication

**Review by jbruestle — COMMENTED** [review]
Seems good to me minus my one comment, adding Frank for additional visibility.

File: pmlc/target/intel_gen_ocl_spirv/spirv_target.cc (hunk 1 above)
Comment (adstraw) [reply pmlc/target/intel_gen_ocl_spirv/spirv_target.cc:27]: I feel like this is OK.  The `auto version = ` can signify the default (latest) version and we can add any other versions needed, currently just 1_2, in the conditional statements below.  Not sure how to make it more general, either.  I don't think there's a way to pass `spriv::Version` through the PassOptions directly.  Meaning, we have to translate from int (or string) to `spriv::Version` at some level.  Meaning, we need this case logic at some level.

File: pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc (hunk 3 above)
Comment (adstraw) [reply pmlc/conversion/gpu_to_spirv/gpu_to_spirv.cc]: Done.

File: pmlc/target/x86/register.cc (hunk 2 above)
Comment (adstraw) [reply pmlc/target/x86/register.cc]: Nope, removed.

File: pmlc/compiler/program.cc [hunk 4]
```diff
@@ -149,8 +149,19 @@ void Program::compile(StringRef targetName, bool collectPasses,
...
+  auto begOpts = targetNameAndOptions.find('{');
+  auto targetName = targetNameAndOptions.substr(0, begOpts);
+  auto targetOptions = targetNameAndOptions.substr(begOpts);
+
+  // if target options are specified
+  if (!targetOptions.empty()) {
+    // trim off curly braces
+    auto endOpts = targetOptions.find('}');
+    targetOptions = targetOptions.substr(1, endOpts - 1);
+  }
```
Comment (flaub) [inline pmlc/compiler/program.cc:161]: This is the only part that gives me pause. I wonder if we can use the existing parse pipeline functionality of MLIR for this.

**Comment by flaub** [comment]
LGTM, but we might want to come up with a more 'MLIR' way to parse the pipeline options. When I have some cycles I can investigate this, but I don't think it should block landing this.

File: pmlc/compiler/program.cc (hunk 4 above)
Comment (adstraw) [reply pmlc/compiler/program.cc:161]: This code separates target name from options.  E.g. PLAIDML_TARGET="mytarget{myoption1=true myoption2=7}" you get two strings targetName = "mytarget" and targetOptions = "myoption1=true myoption2=7".  I did not find an MLIR utility that does this.  I do parse the targetOptions with an MLIR utility called `createFromString`.

**Review by jbruestle — APPROVED** [review]
Looks good!

**Pull request merged**